src/resolvers/directives/requireOneOfRoles.ts

```diff
@@ -1,33 +1,49 @@
 import { DirectiveResolverMap } from '../../typings/resolverMap';
 import { oneLine } from 'common-tags';
 import { ApiError } from '../../helpers/apiError';
+import { resolvers as requireAuthResolvers } from './requireAuth';
+
+const { requireAuth } = requireAuthResolvers;
 
 export const resolvers: Pick<DirectiveResolverMap, 'requireOneOfRoles'> = {
-  async requireOneOfRoles(next, _source, { roles }, { viewer }) {
-    if (!viewer) {
-      throw new ApiError('NOT_AUTHENTICATED');
-    }
+  requireOneOfRoles: async (next, source, { roles }, context, info) => {
+    const checkForMatchingRoles = async () => {
+      const { viewer } = context;
+      if (!viewer) {
+        throw new ApiError('NOT_AUTHENTICATED');
+      }
+
+      if (viewer.role === null) {
+        throw new ApiError('NOT_AUTHORIZED');
+      }
 
-    if (viewer.role === null) {
-      throw new ApiError('NOT_AUTHORIZED');
-    }
+      if (viewer.role && roles.includes(viewer.role)) {
+        return next();
+      }
 
-    if (viewer.role && roles.includes(viewer.role)) {
-      return next();
-    }
+      let errorMessage: string;
+      if (roles.length === 1) {
+        errorMessage = `Only users with an "${
+          roles[0]
+        }" role can perform this operation`;
+      } else {
+        errorMessage = oneLine`
+          Only users with one of the following
+          roles are allowed to perform this operation: ${roles.join(', ')}
+        `;
+      }
 
-    let errorMessage: string;
-    if (roles.length === 1) {
-      errorMessage = `Only users with an "${
-        roles[0]
-      }" role can perform this operation`;
-    } else {
-      errorMessage = oneLine`
-        Only users with one of the following
-        roles are allowed to perform this operation: ${roles.join(', ')}
-      `;
-    }
+      throw new ApiError('ROLE_MISMATCH', errorMessage);
```

Comment:
Isn't this technically a `NOT_AUTHORIZED` error?

Reply:
Yeah but the error code and message would be more helpful for users who have a role but the role does not satisfy what's needed to perform the operation. For regular users who do not have roles, `NOT_AUTHORIZED` is used.